transports/webrtc/src/udp_mux.rs

```diff
@@ -0,0 +1,382 @@
+            conn: Box::new(conn),
+        }
+    }
+}
+
+/// An event emitted by [`UDPMuxNewAddr`] when it's polled.
+#[derive(Debug)]
+pub enum UDPMuxEvent {
+    /// Connection error. UDP mux should be stopped.
+    Error(Error),
+    /// Got a [`NewAddr`] from the socket.
+    NewAddr(NewAddr),
+    /// Non-important event. Can be ignored.
+    None,
+}
+
+/// A modified version of [`webrtc_ice::udp_mux::UDPMuxDefault`], which reports previously unseen
+/// addresses instead of ignoring them.
+pub struct UDPMuxNewAddr {
+    /// The params this instance is configured with.
+    /// Contains the underlying UDP socket in use
+    params: UDPMuxParams,
+
+    /// Maps from ufrag to the underlying connection.
+    conns: Mutex<HashMap<String, UDPMuxConn>>,
+
+    /// Maps from socket address to the underlying connection.
+    address_map: RwLock<HashMap<SocketAddr, UDPMuxConn>>,
+
+    /// Set of the new addresses to avoid sending the same address multiple times.
+    new_addrs: RwLock<HashSet<SocketAddr>>,
+
+    /// `true` when UDP mux is closed.
+    is_closed: AtomicBool,
+}
+
+impl UDPMuxNewAddr {
+    /// Creates a new UDP muxer.
+    pub fn new(params: UDPMuxParams) -> Arc<Self> {
+        Arc::new(Self {
+            params,
+            conns: Mutex::default(),
+            address_map: RwLock::default(),
+            new_addrs: RwLock::default(),
+            is_closed: AtomicBool::new(false),
+        })
+    }
+
+    /// Create a muxed connection for a given ufrag.
+    async fn create_muxed_conn(self: &Arc<Self>, ufrag: &str) -> Result<UDPMuxConn, Error> {
+        let local_addr = self.params.conn.local_addr().await?;
+
+        let params = UDPMuxConnParams {
+            local_addr,
+            key: ufrag.into(),
+            udp_mux: Arc::downgrade(self) as Weak<dyn UDPMuxWriter + Sync + Send>,
+        };
+
+        Ok(UDPMuxConn::new(params))
+    }
+
+    /// Returns a muxed connection if the `ufrag` from the given STUN message matches an existing
+    /// connection.
+    async fn conn_from_stun_message(&self, buffer: &[u8], addr: &SocketAddr) -> Option<UDPMuxConn> {
+        match ufrag_from_stun_message(buffer, true) {
+            Ok(ufrag) => {
+                let conns = self.conns.lock().await;
+                conns.get(&ufrag).map(Clone::clone)
+            }
+            Err(e) => {
+                log::debug!("{} (addr={})", e, addr);
+                None
+            }
+        }
+    }
+
+    /// Returns true if the UDP muxer is closed.
+    pub fn is_closed(&self) -> bool {
+        return self.is_closed.load(Ordering::Relaxed);
+    }
+
+    /// Reads from the underlying UDP socket and either reports a new address or proxies data to the
+    /// muxed connection.
+    pub async fn read_from_conn(&self) -> UDPMuxEvent {
+        // TODO: avoid reallocating the buffer
+        let mut buffer = [0u8; RECEIVE_MTU];
+        let conn = &self.params.conn;
+
+        let res = conn.recv_from(&mut buffer).await;
```

Comment:
I know this is mostly copied but I think if we already copy it, we might as well make things better for us and change the way this is implemented?

Reading bytes from a connection and turning them into messages f.e. is something where `asynchronous-codec` could help perhaps.